core/src/main/java/org/apache/iceberg/DeleteFileIndex.java

```diff
@@ -76,7 +76,7 @@
     ImmutableMap.Builder<Integer, Types.StructType> builder = ImmutableMap.builder();
     specsById.forEach((specId, spec) -> builder.put(specId, spec.partitionType()));
     this.partitionTypeById = builder.build();
-    this.wrapperById = Maps.newHashMap();
+    this.wrapperById = Maps.newConcurrentMap();
```

Comment:
My only question on this PR is whether we should switch to using a Caffeine loading cache here? I just notice in some places we do that, and in others we just use the concurrent map.

Reply:
Is there an advantage to using a Caffeine cache here? I think it most cases, we use a cache when we need more features than just `computeIfAbsent`, like invalidation and expiration. We should probably prefer using standard JVM classes to libraries like Caffeine where possible.

Reply:
My only preference is that it lets you set the "ifAbsent" method as default, so you don't actually have to call computeIfAbsent when getting. But I have no preference, just the usage in ManifestGroup doesn't use invalidation or expiration so I was wondering if there was a reason.

Reply:
The one quirk to be aware of is that `computeIfAbsent` is pessimistic whereas Caffeine's is optimistic. This doesn't mean to prefer it over standard JDK classes, but does that boilerplate optimization which may matter if a hotspot.

CHM in v8 finds the hashbin, locks it, searches for the entry, and compute if absent. If the entry is found then it has the cost of a `putIfAbsent`, which similarly blocks on writes. In v9+ the compute method optimistically avoids locking if the first bin entry is the desired one but if not then it locks and searches. This has a modest gain, but still overhead for hot keys.

As Caffeine is a cache, it leans towards the entry being present. Therefore it does an `get`, validates the entry if found (e.g. expiration), and fallback to a `compute` if absent. This means that a cache hit is lock-free while a miss is slightly penalized, but by a few nanos as it will most often do expensive I/O so insignificant.

You can see this behavior for yourself in the [benchmarks](https://github.com/ben-manes/caffeine/wiki/Benchmarks#compute). So while no reason to use Caffeine, it's a subtle quirk to be aware of in your own usages of CHM.

Reply:
Thanks for the info! That definitely the case for this usage since it will be set once read many. I'll definitely consider this in future cache/map situations

Reply:
Thanks for the great context, @ben-manes!

I still expect IO to dominate the cost of planning, but we should certainly keep this behavior in mind in the future. I'd be for changing this (and possibly other places) over to a cache, but let's keep this PR focused since it fixes the current flaky test bug.

Reply:
Just to clarify you can do the optimization directly, but may want to comment as it first appears unnecessary.

```java
V value = map.get(key);
if (value == null) {
  value = map.computeIfAbsent(key, mappingFunction);
}
return value;
```